lib/uri/common.rb

```diff
@@ -62,22 +64,38 @@ def make_components_hash(klass, array_hash)
 
   include REGEXP
 
-  @@schemes = {}
+  SCHEME_LIST_MUTEX = Mutex.new
+  private_constant :SCHEME_LIST_MUTEX
+
   # Returns a Hash of the defined schemes.
+  # The list is lazily calculated.
   def self.scheme_list
-    @@schemes
+    return const_get(:SCHEMES) if defined?(SCHEMES)
+
+    SCHEME_LIST_MUTEX.synchronize do
+      const_set(:SCHEMES, ObjectSpace.
+        each_object(Class).
+        select { |klass| klass < URI::Generic }.
+        each_with_object({}) { |klass, acc| acc[klass.name.split('::').last.upcase] = klass }.
+        freeze)
+    end
+  end
+
+  # Re-calculate scheme list
+  def self.refresh_scheme_list
```

Comment:
This is yet another hack, now the API exposes the caching to the end user, that's leaking implementation details.
It's ugly, there is no other word.